pkg/natdiscovery/natdiscovery.go

```diff
@@ -0,0 +1,188 @@
+	serverPort      int
+}
+
+func New(localEndpoint *types.SubmarinerEndpoint) (Interface, error) {
+	return newNatDiscovery(localEndpoint)
+}
+
+func newNatDiscovery(localEndpoint *types.SubmarinerEndpoint) (*natDiscovery, error) {
+	port, err := extractNATDiscoveryPort(localEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	requestCounter, err := randomRequestCounter()
+	if err != nil {
+		return nil, err
+	}
+
+	return &natDiscovery{
+		localEndpoint:   localEndpoint,
+		serverPort:      port,
+		remoteEndpoints: map[string]*remoteEndpointNAT{},
+		findSrcIP:       findPreferredSourceIP,
+		requestCounter:  requestCounter,
+	}, nil
+}
+
+func randomRequestCounter() (uint64, error) {
+	n, err := rand.Int(rand.Reader, big.NewInt(1000))
+	if err != nil {
+		return 0, errors.Wrapf(err, "generatimg random request counter")
+	}
+
+	return n.Uint64(), nil
+}
+
+var errorNoNatDiscoveryPort = errors.New("no nat discovery port in endpoint")
+
+func extractNATDiscoveryPort(endpoint *types.SubmarinerEndpoint) (int, error) {
+	if endpoint.Spec.PortConfig == nil {
+		return 0, errorNoNatDiscoveryPort
+	}
+
+	if portString, ok := endpoint.Spec.PortConfig[v1.NATDiscoveryPortConfig]; ok {
+		port, err := strconv.Atoi(portString)
+		if err != nil {
+			return 0, errors.Wrapf(err, "error trying to extract port number from %q in %s",
+				portString, v1.NATDiscoveryPortConfig)
+		}
+
+		return port, nil
+	}
+
+	return 0, errorNoNatDiscoveryPort
+}
+
+func (nd *natDiscovery) Run(stopCh <-chan struct{}, natResultCh chan remoteEndpointNAT) error {
+	err := nd.runListener(stopCh)
+	if err != nil {
+		return err
+	}
+
+	go wait.Until(func() {
+		nd.checkEndpointList()
+	}, time.Second, stopCh)
+
+	return nil
+}
+
+func (nd *natDiscovery) AddEndpoint(endpoint *types.SubmarinerEndpoint) {
+	remoteNAT := newRemoteEndpointNAT(endpoint)
+
+	// support a remote cluster endpoint which still hasn't implemented this protocol
+	if _, err := extractNATDiscoveryPort(endpoint); err == errorNoNatDiscoveryPort {
+		remoteNAT.useLegacyNATSettings()
+	}
+
+	nd.Lock()
+	defer nd.Unlock()
+
+	nd.remoteEndpoints[endpoint.Spec.CableName] = remoteNAT
+}
+
+func (nd *natDiscovery) RemoveEndpoint(endpointName string) {
+	nd.Lock()
+	defer nd.Unlock()
+	delete(nd.remoteEndpoints, endpointName)
+}
+
+func (nd *natDiscovery) checkEndpointList() {
+	nd.Lock()
+	defer nd.Unlock()
+
+	for _, endpointNAT := range nd.remoteEndpoints {
+		if endpointNAT.hasTimedOut() {
+			endpointNAT.useLegacyNATSettings()
+		} else if endpointNAT.shouldCheck() {
+			err := nd.sendCheckRequest(endpointNAT)
+			if err != nil {
+				klog.Errorf("error sending check request to %q: %s", endpointNAT.Endpoint.Spec.CableName, err)
+			}
+		}
+	}
+}
+
+var UnknownEndpoint = errors.New("unknown endpoint")
+
+func (nd *natDiscovery) getRemoteEndpointNAT(endpointName string) *remoteEndpointNAT {
+	nd.Lock()
+	defer nd.Unlock()
+
+	remoteNat, ok := nd.remoteEndpoints[endpointName]
+	if ok {
+		return remoteNat
+	}
+
```

Comment:
So access to the map is protected by the lock but the `remoteEndpointNAT` instance is unprotected once returned which could be problematic.  So everywhere we access the `remoteEndpoints` map and the mutable fields in `remoteEndpointNAT` needs to be protected. But ideally we want to avoid too wide a lock. We could use a `sync.Map` for `remoteEndpoints` and add a lock to `remoteEndpointNAT` to protect access to its mutable fields.